lib/mli_parser.ml

```diff
@@ -1,207 +1,128 @@
 module Code_block = struct
-  type metadata = {
-    language_tag : string Odoc_parser.Loc.with_location;
-    labels : string Odoc_parser.Loc.with_location option;
-  }
+  type metadata = { language_tag : string; labels : string option }
 
   type t = {
-    location : Odoc_parser.Loc.span;
     metadata : metadata option;
-    contents : string;
+    content : Location.t; (* Location of the content *)
+    code_block : Location.t; (* Location of the enclosing code block *)
   }
 end
 
-(* odoc-parser adjusts for the initial [** *)
-let docstring_start_adjustment = String.length "(**"
-
-let drop_last lst =
-  match List.rev lst with
-  | [] -> None
-  | last :: rev_tl -> Some (List.rev rev_tl, last)
+(* Parse and extract code block metadata from an odoc formatted docstring.
 
-(* drop_first_and_last [1; 2; 3; 4] = Some (1, Some ([2; 3], 4)). *)
-let drop_first_and_last = function
-  | [] -> None
-  | first :: tl -> Some (first, drop_last tl)
+   Code blocks are the only thing we're interested in. This function parses
+   the given text and extracts the metadata and enough location information
+   from the code blocks be able to String.sub them out of the original text.
 
-let slice lines ~(start : Odoc_parser.Loc.point) ~(end_ : Odoc_parser.Loc.point)
-    =
-  let lines_to_include =
-    Util.Array.slice lines ~from:(start.line - 1) ~to_:(end_.line - 1)
-    |> Array.to_list
-  in
-  match drop_first_and_last lines_to_include with
-  | None -> ""
-  | Some (line, None) ->
-      String.sub line start.column (end_.column - start.column)
-  (* Imagine we were slicing the file from (Line 2, Column 3) to (Line 6, Column 7):
+   [location] is the location of this docstring within the original file
+   (ie, the location of the contents of the documentation comment). This is
+   required so we can splice out the code blocks from the original file.
 
-       0123456789
-       1 ----------
-       2 ---[---
-       3 ---------
-       4 --
-       5 ----------
-       6 -------]--
-       7 ----------
-       8 ----------
+   The results are prepended in reverse order onto [acc]. *)
+let extract_code_block_info acc ~(location : Lexing.position) ~docstring =
+  let open Odoc_parser in
```

Comment:
How bad would the code look without that `open`? I find code using `open` often hard to reason about and try to avoid it if possible.

Reply:
It just looked a little noisy. I was definitely against `open`s in the past, but these days I find merlin really helps disambiguation. I kept it local to the function that requires it, but if you're opposed I'll take it out.

Reply:
Removing the `open Code_block` was not a problem, happy to do that. Removing the `open Odoc_parser` led to a lot of additional noise, IMO. Here's a commit with the changes: https://github.com/jonludlam/mdx/commit/dcce65ec579364b9eef0d48493252deaf2a6afd2